fem/ceed/algebraic.cpp

```diff
@@ -0,0 +1,989 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "algebraic.hpp"
+
+#include "../bilinearform.hpp"
+#include "../fespace.hpp"
+#include "../ceed/solvers-atpmg.hpp"
+#include "../ceed/full-assembly.hpp"
+#include "../pfespace.hpp"
+#include "../../general/forall.hpp"
+
+namespace mfem
+{
+
+namespace ceed
+{
+
+#ifdef MFEM_USE_CEED
+
+/** Construct a ceed::Operator from a CeedOperator */
+class WrapOperator : public ceed::Operator
+{
+public:
+   /** This object takes ownership of ceed_op and will delete it */
+   WrapOperator(CeedOperator ceed_op)
+   {
+      oper = ceed_op;
+      CeedElemRestriction er;
+      CeedOperatorGetActiveElemRestriction(oper, &er);
+      int s;
+      CeedElemRestrictionGetLVectorSize(er, &s);
+      height = width = s;
+      CeedVectorCreate(internal::ceed, height, &v);
+      CeedVectorCreate(internal::ceed, width, &u);
+   }
+
+   mfem::Operator * SetupRAP(const mfem::Operator *Pi, const mfem::Operator *Po)
+   {
+      return Operator::SetupRAP(Pi, Po);
+   }
+};
+
+/** Wraps a CeedOperator in an mfem::Operator, with essential boundary
+    conditions and a prolongation operator for parallel application. */
+class ConstrainedMFEMCeedOperator : public mfem::Operator
```

Comment:
Rename to `ConstrainedOperator`.

Reply:
This class should also be move with `ceed/operator.{hpp,cpp}`.

Reply:
I renamed it, but I didn't move it since I'm not sure it should be user-facing. Right now it is just helper code for the algebraic solver in the .cpp file.

Reply:
What would not make it good enough to be user-facing?

Reply:
How many users want to take a `ceed::Operator`, an `Array` of essential boundary dofs, and an `Operator` representing the parallel prolongation operator to make an `mfem::Operator`? I just don't see this as a useful general class.